### PR TITLE
feat: As a Module creator, I can publish my Module

### LINKED ISSUE #20: [US] As a Module Developer, I can publish my Module
As a module creator, I want to publish my modules on the Module Registry so that they can be used in Portals

## Requirements
* Anyone can create a Module. No rule to prevent anyone from creating a Module
* Modules are registered through the `Module Registry` smart contract
* Modules are identified via their Smart Contract addresses
* Modules published should be stored in an array (for discoverability)
* Modules' metadata should be stored in a mapping when the module is published
    * The module smart contract address is used as a key
    * Module name and description (optional) are stored inside the mapping (struct)
* Publishing a module should trigger an event

## Todo
- [ ] Create a `ModuleRegistry` contract
- [ ] Create a `Module` struct to store a Module metadata
- [ ] Add a mapping to store modules (`address => Module`)
- [ ] Add an array to store modules
- [ ] Add a `register()` function to add a new Module to the mapping and the array
- [ ] To register a Module, it must have a name and an address, throw custom errors if not
- [ ] When a Module is registered, emit a custom `ModuleRegistered` event

- [ ] Add a `ModuleRegistry` script to be able to deploy it easily + a pnpm command to trigger it
- [ ] Add a `ModuleRegistry` set of unit tests to cover the happy path + the errors and events triggering

## Acceptance criteria
* Should include a script to deploy the `Module Registry`
* Should include 100% unit-tests (including accessing the metadata through the mapping)

## Additional information
| Workload estimate | 
|--------|
| TBD |

### PR DESCRIPTION
**Pull request by satyajeetkolhapure**
### Related ticket

Fixes #20 

### Type of change

- [ ] Chore
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update

## Check list

- [x] Unit tests for any smart contract change
- [x] Contracts and functions are documented

### CODE REVIEW

File: src/ModuleRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+   */
+  function isContractAddress(address contractAddress) public view returns (bool) {
+    return contractAddress.code.length > 0;
+  }
+
+  /**
+   * @notice Get the administrator address for this contract
+   * @return The contract's administrator address
+   */
+  function getAdmin() public view returns (address) {
```
Comment (alainncls) [inline src/ModuleRegistry.sol:54]: To remove

File: src/ModuleRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ * @author Consensys
+ * @notice This contract aims to manage the Modules used by the Portals, including their discoverability
+ */
+contract ModuleRegistry is Initializable {
+  /// @dev The contract administrator
+  address private admin;
+  /// @dev The list of Modules, accessed by their ID
+  mapping(address id => Module module) public modules;
+  /// @dev The number of Modules already registered
+  uint256 public numberOfModules = 0;
```
Comment (alainncls) [inline src/ModuleRegistry.sol:19]: No need, in #41 we'll add an array of modules address and read its length to get this value

File: src/ModuleRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      revert ModuleInvalid();
+    }
+
+    // Module address is used to identify uniqueness of the module
+    if (bytes(modules[moduleAddress].name).length > 0) {
+      revert ModuleAlreadyExists();
+    }
+
+    modules[moduleAddress] = Module(name, description, moduleAddress);
+    numberOfModules++;
```
Comment (alainncls) [inline src/ModuleRegistry.sol:88]: To remove

File: src/ModuleRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    emit ModuleCreated(name, description, moduleAddress);
+  }
+}
+
+//TODO: Replace this interface with proper module interface in future user story
+interface ModuleInterface {
```
Comment (alainncls) [inline src/ModuleRegistry.sol:94]: Let's extract this interface in `src/interface` even if it's not final

File: test/ModuleRegistry.t.sol [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  function testCannotCreateModuleWithoutName() public {
+    vm.expectRevert(ModuleRegistry.ModuleNameMissing.selector);
+    moduleRegistry.createModule("", expectedDescription, expectedAddress);
+  }
+
+  function testCannotCreateModuleWithInvalidModuleAddress() public {
+    vm.expectRevert(ModuleRegistry.ModuleAddressInvalid.selector);
+    moduleRegistry.createModule(expectedName, expectedDescription, vm.addr(1));
```
Comment (alainncls) [inline test/ModuleRegistry.t.sol:42]: Can you add a comment to explain that `vm.addr(1)` is giving an EOA please?

File: script/ModuleRegistry.s.sol [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.21;
+
+import { Script } from "forge-std/Script.sol";
+import { ModuleRegistry } from "../src/ModuleRegistry.sol";
+
+contract ModuleRegistryScript is Script {
```
Comment (alainncls) [inline script/ModuleRegistry.s.sol:7]: Can you add a script in the `package.json` to call this deployment from a pnpm command?
FYI, we already have a `deploy:SchemaRegistry:goerli` command 😉

**Review by alainncls — CHANGES_REQUESTED** [review]

File: src/ModuleRegistry.sol [hunk 7]
```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.21;
+
+import { Module } from "./struct/Module.sol";
+import "openzeppelin-contracts/contracts/proxy/utils/Initializable.sol";
+import "openzeppelin-contracts/contracts/utils/introspection/ERC165Checker.sol";
```
[the remark this answers is not in the document]
Comment (satyajeetkolhapure) [reply src/ModuleRegistry.sol]: fixed

**Review by alainncls — APPROVED** [review]

**Pull request merged**